src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -48,5 +50,129 @@ public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private Dictionary<Tuple<string, string>, EventBase> EventsInProgress = new Dictionary<Tuple<string, string>, EventBase>();
+
+        private Dictionary<string, List<EventBase>> CompletedEvents = new Dictionary<string, List<EventBase>>();
+
+        internal string GenerateNewRequestId()
+        {
+            return Guid.NewGuid().ToString();
+        }
+
+        internal void StartEvent(string requestId, EventBase eventToStart)
+        {
+            EventsInProgress[new Tuple<string, string>(requestId, eventToStart["event_name"])] = eventToStart;
+        }
+
+        internal void StopEvent(string requestId, EventBase eventToStop)
+        {
+            Tuple<string, string> eventKey = new Tuple<string, string>(requestId, eventToStop["event_name"]);
+
+            // Locate the same name event in the EventsInProgress map
+            EventBase eventStarted = null;
+            if (EventsInProgress.ContainsKey(eventKey))
+            {
+                eventStarted = EventsInProgress[eventKey];
+            }
+
+            // If we did not get anything back from the dictionary, most likely its a bug that StopEvent
+            // was called without a corresponding StartEvent
+            if (null == eventStarted)
```

Comment:
Yeah we could probably do that. Actually I had some further fine-tuning ideas in mind when I coded this (<sub>and I'm not even referring to those radical ideas 2 weeks ago</sub>) ... but wait, I thought we decided to mimic pre-existing implementation in order to save time and effort from both you and me? ;-)

BTW, another one (probably two) telemetry event PRs are on the way. Let's see what we can accomplish in Friday.